tfconfig/provider_ref.go

```diff
@@ -38,46 +35,73 @@ func decodeRequiredProvidersBlock(block *hcl.Block) (map[string]*ProviderRequire
 					VersionConstraints: []string{version},
 				}
 			}
+			continue
+		}
+
+		kvs, mapDiags := hcl.ExprMap(attr.Expr)
+		if mapDiags.HasErrors() {
+			diags = append(diags, &hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid required_providers object",
+				Detail:   "Required providers entries must be strings or objects.",
+				Subject:  attr.Expr.Range().Ptr(),
+			})
+			continue
+		}
+
+		var pr ProviderRequirement
+
+		for _, kv := range kvs {
+			key, keyDiags := kv.Key.Value(nil)
+			if keyDiags.HasErrors() {
+				diags = append(diags, keyDiags...)
+				continue
+			}
 
-		case expr.Type().IsObjectType():
-			var pr ProviderRequirement
-			if expr.Type().HasAttribute("version") {
-				var version string
-				err := gocty.FromCtyValue(expr.GetAttr("version"), &version)
-				if err == nil {
-					pr.VersionConstraints = append(pr.VersionConstraints, version)
-				} else {
+			if key.Type() != cty.String {
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Invalid Attribute",
```

Comment:
I would drop this error, or at least make it a warning instead (though I'm not sure if we have many warnings, so that might not be a good idea either). 

We generally ignore unexpected fields as part of the (vague, hand-wavey) future-proofing. `terraform-config-inspect` is intended to be a very permissive tool that only knows what it knows directly and ignores anything extra.

Reply:
I put this in since it was in the category of "should never happen", and would have generated a parsing error before as well. Come to think of it though, I'm not sure if we could ever get here without a string at all 🤔

Reply:
Looking at `hcl.ExprMap`, the returned values are entirely up to the expression implementation, however they should only be returning a value at all if there this is a valid map, which requires string keys/attributes. We can probably just drop this error altogether.

Reply:
I would generally lean also towards permissiveness in the behaviour, esp. because we also use it in the language server, where we expect to be parsing invalid configs pretty often.

That said as long as we continue parsing the rest of the config (which seems to be the case here) and not "stop & return nil and diags", I'm ok with that as diags can always be dealt with.

Reply:
Just to note, the old code would have returned an error if this is even possible at some level. The string check is basically a failsafe in case a faulty `ExprMap` implementation returns something incorrect for the key value, but maybe that's overkill here and we should just let it panic in that case.

Reply:
Er .... looking at this again I just misread the error case :(